README.md
# Apparition - A Chrome driver for Capybara #

[![Build Status](https://secure.travis-ci.org/twalpole/apparition.svg)](http://travis-ci.org/twalpole/apparition)

Apparition is a driver for [Capybara](https://github.com/teamcapybara/capybara). It allows you to
run your Capybara tests in the Chrome browser via CDP (no selenium or chromedriver needed) in a headless or
headed configuration. It started as a fork of Poltergeist and attempts to maintain as much compatibility
with the Poltergeist API as possible.  Implementing the `capybara-webkit` specific driver methods has also begun.

## Getting help ##

Questions should be posted [on Stack
Overflow, using the 'capybara' tag](http://stackoverflow.com/questions/tagged/capybara) and mentioning
you are using the apparition driver.

Bug reports should be posted [on GitHub](https://github.com/twalpole/apparition/issues) (and be sure
to read the bug reporting guidance below).

## Installation ##

Add either

``` ruby
gem 'apparition'
```

or

``` ruby
gem 'apparition', github: 'twalpole/apparition'
```

to your Gemfile and run `bundle install`.

In your test setup add:

``` ruby
require 'capybara/apparition'
Capybara.javascript_driver = :apparition
```

If you were previously using the `:rack_test` driver, be aware that your app will now run in a separate thread and this can have
consequences for transactional tests. [See the Capybara README for more detail](https://github.com/teamcapybara/capybara/blob/master/README.md#transactions-and-database-setup).

## What's supported? ##

Apparition supports all Capybara features, and the following extended features:

* `page.status_code`
* `page.response_headers`
* `page.driver.render_base64(format, options)`
* `page.driver.scroll_to(left, top)`
* `page.driver.basic_authorize(user, password)`
* `page.driver.set_proxy(host, port, type, user, password, bypass: [bypass list passed to chrome])`
* cookie handling
* extra headers

There are some additional features:

### Taking screenshots with some extensions ###

You can grab screenshots of the page at any point by calling
`save_screenshot('/path/to/file.png')`.

By default, only the viewport will be rendered (the part of the page that is in
view). To render the entire page, use `save_screenshot('/path/to/file.png',
full: true)`.

You also have an ability to render selected element. Pass option `selector` with
any valid CSS element selector to make a screenshot bounded by that element
`save_screenshot('/path/to/file.png', selector: '#id')`.

If the desired image format is not identifiable from the filename passed you can
also pass in a `format:` option with accepable values being `:png` or `:jpeg`

If, for some reason, you need a base64 encoded screenshot you can simply call
`render_base64` which will return your encoded image. Additional options are the
same as for `save_screenshot`.

### Clicking precise coordinates ###

Sometimes its desirable to click a very specific area of the screen. You can accomplish this with
`page.driver.click(x, y)`, where x and y are the screen coordinates.

### Remote debugging (not yet implemented) ###

If you use the `:inspector => true` option (see below), remote debugging
will be enabled.

When this option is enabled, you can insert `page.driver.debug` into
your tests to pause the test and launch a browser which gives you the
WebKit inspector to view your test run with.

You can register this debugger driver with a different name and set it
as the current javascript driver. By example, in your helper file:

```ruby
Capybara.register_driver :apparition_debug do |app|
  Capybara::Apparition::Driver.new(app, :inspector => true)
end
# Capybara.javascript_driver = :apparition
Capybara.javascript_driver = :apparition_debug
```

[Read more
here](https://www.jonathanleighton.com/articles/2012/poltergeist-0-6-0/)

### Manipulating request headers ###

You can manipulate HTTP request headers with these methods:

``` ruby
page.driver.headers # => {}
page.driver.headers = { "User-Agent" => "Apparition" }
page.driver.add_headers("Referer" => "https://example.com")
page.driver.headers # => { "User-Agent" => "Apparition", "Referer" => "https://example.com" }
```

Notice that `headers=` will overwrite already set headers. You should use
`add_headers` if you want to add a few more. These headers will apply to all
subsequent HTTP requests (including requests for assets, AJAX, etc). They will
be automatically cleared at the end of the test. You have ability to set headers
only for the initial request:

``` ruby
page.driver.headers = { "User-Agent" => "Apparition" }
page.driver.add_header("Referer", "http://example.com", permanent: false)
page.driver.headers # => { "User-Agent" => "Apparition", "Referer" => "http://example.com" }
visit(login_path)
page.driver.headers # => { "User-Agent" => "Apparition" }
```

This way your temporary headers will be sent only for the initial request, and related 30x redirects. All
subsequent request will only contain your permanent headers. If the temporary
headers should not be sent on related 30x redirects, specify `permanent: :no_redirect`.

### Inspecting network traffic ###

You can inspect the network traffic (i.e. what resources have been
loaded) on the current page by calling `page.driver.network_traffic`.
This returns an array of request objects. A request object has a
`response_parts` method containing data about the response chunks.

You can inspect requests that were blocked by a whitelist or blacklist
by calling `page.driver.network_traffic(:blocked)`. This returns an array of
request objects. The `response_parts` portion of these requests will always
be empty.

Please note that network traffic is not cleared when you visit new page.
You can manually clear the network traffic by calling `page.driver.clear_network_traffic`
or `page.driver.reset`

### Manipulating cookies ###

The following methods are used to inspect and manipulate cookies:

* `page.driver.cookies` - a hash of cookies accessible to the current
  page. The keys are cookie names. The values are `Cookie` objects, with
  the following methods: `name`, `value`, `domain`, `path`, `secure?`,
  `httponly?`, `samesite`, `expires`.
* `page.driver.set_cookie(name, value, options = {})` - set a cookie.
  The options hash can take the following keys: `:domain`, `:path`,
  `:secure`, `:httponly`, `:samesite`, `:expires`. `:expires` should be a
  `Time` object.
* `page.driver.remove_cookie(name)` - remove a cookie
* `page.driver.clear_cookies` - clear all cookies

## Customization ##

You can customize the way that Capybara sets up Apparition via the following code in your
test setup:

``` ruby
Capybara.register_driver :apparition do |app|
  Capybara::Apparition::Driver.new(app, options)
end
```

`options` is a hash of options. The following options are supported:

*   `:headless` (Boolean) - When false, run the browser visibly
*   `:remote` (Boolean) - When true, connect to remote browser instead of starting locally (see [below](#Remote Chrome Driver))
*   `:debug` (Boolean) - When true, debug output is logged to `STDERR`.
*   `:logger` (Ruby logger object or any object responding to `puts`) - When present, debug output is written to this object
*   `:browser_logger` (`IO` object) - This is where your `console.log` statements will show up. Default: `STDOUT`
*   `:timeout` (Numeric) - The number of seconds we'll wait for a response
    when communicating with Chrome. Default is 30.
*   `:inspector` (Boolean, String) - See 'Remote Debugging', above.
*   `:js_errors` (Boolean) - When false, JavaScript errors do not get re-raised in Ruby.
*   `:window_size` (Array) - The dimensions of the browser window in which to test, expressed
    as a 2-element array, e.g. [1024, 768]. Default: [1024, 768]
*   `:screen_size` (Array) - The dimensions the window size will be set to when Window#maximize is called in headless mode.  Expressed
    as a 2-element array, e.g. [1600, 1200]. Default: [1366, 768]
*   `:extensions` (Array) - An array of JS files to be preloaded into
    the browser. Useful for faking or mocking APIs.
*   `:url_blacklist` (Array) - Default session url blacklist - expressed as an array of strings to match against requested URLs.
*   `:url_whitelist` (Array) - Default session url whitelist - expressed as an array of strings to match against requested URLs.
*   `:ignore_https_errors` (Boolean) - Ignore certificate errors when connecting to https URLs.
*   `:browser_options` (Hash) - Extra command line options to pass to Chrome when starting
*   `:skip_image_loading` (Boolean) - Don't load images

### Remote Chrome Driver ###
Apparition can connect to already running instance of chrome. 
Remote mode is useful when running tests in CI and chrome is available as separate docker container.

In order to use remote browser - set up apparition in the following way:
```ruby
Capybara.register_driver :apparition do |app|
  browser_options = { 'remote-debugging-address' => '127.0.0.1', 'remote-debugging-port' => 9999 }
  Capybara::Apparition::Driver.new(app, remote: true, browser_options: browser_options)
end
```

### URL Blacklisting & Whitelisting ###
Apparition supports URL blacklisting, which allows you
to prevent scripts from running on designated domains:

```ruby
page.driver.browser.url_blacklist = ['http://www.example.com']
```

and also URL whitelisting, which allows scripts to only run
on designated domains:

```ruby
page.driver.browser.url_whitelist = ['http://www.example.com']
```

If you are experiencing slower run times, consider creating a
URL whitelist of domains that are essential or a blacklist of
domains that are not essential, such as ad networks or analytics,
to your testing environment.

### Emulating a mobile device ###

Desktop Chrome often won't let you set the window width narrower than 500px
without also enabling mobile emulation, even when Chrome appears to accept the
width.

To test with mobile widths, it's necessary to enable mobile emulation via
CDP with all the required parameters.

```ruby
nexus_5_metrics = {
  mobile: true,
  screenWidth: 412,
  screenHeight: 660,
  width: 412,
  height: 660,
  positionX: 0,
  positionY: 0,

  scale: 1,
  deviceScaleFactor: 2.625,
  screenOrientation: {
    angle: 0,
    type: "portraitPrimary",
  },
}

page.driver.browser.current_page.command(
  "Emulation.setDeviceMetricsOverride",
  **nexus_5_metrics,
)

page.driver.browser.current_page.command(
  "Emulation.setTouchEmulationEnabled",
  { enabled: true },
)
```

These settings put Chrome in mobile emulation mode in the same fashion as
enabling the emulation mode from the developer tools.

### Timing problems ###

Sometimes tests pass and fail sporadically. This is often because there
is some problem synchronising events properly. It's often
straightforward to verify this by adding `sleep` statements into your
test to allow sufficient time for the page to settle.

If you have these types of problems, read through the [Capybara
documentation on asynchronous
JavaScript](https://github.com/teamcapybara/capybara#asynchronous-javascript-ajax-and-friends)
which explains the tools that Capybara provides for dealing with this.

### Filing a bug ###

If you can provide specific steps to reproduce your problem, or have
specific information that might help track down the problem, then please file a bug on Github.

Include as much information as possible. For example:

* Specific steps to reproduce where possible (failing tests are even
  better)
* The output obtained from running Apparition with `:debug` turned on or ENV['DEBUG'] set
* Screenshots
* Stack traces if there are any Ruby on JavaScript exceptions generated
* The Apparition, Capybara, and Chrome version numbers used
* The operating system name and version used

## Changes ##

Version history and a list of next-release features and fixes can be found in
the [changelog](CHANGELOG.md).

## License ##

Copyright (c) 2019 Thomas Walpole

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Version 0.6.0
Release date: 2020-06-21

### Added

* Support Ruby loggers in the :logger and :browser_logger option [Daniel Orner]
* Support "Remote" Chrome browser [Volodymyr Mykhailyk]
* Support rapid set functionality
* Support click delay
* Experimental: Support sending emoji to text inputs

### Fixed

* User Fetch CDP domain for network interception

### Changed

* Only support Ruby 2.5+

# Version 0.5.0
Release date: 2020-01-26

### Added

* Support :drop_modifiers option in `#drag_to`
* Support setting range input

### Fixed

* Ruby 2.7 keyword arguments warnings
* Error in visibility JS atom [Stefan Wienert]
* Issue with request headers [dabrowt1]

# Version 0.4.0
Release date: 2019-07-15

### Added

* Node#rect added to support spatial filters
* Support for Capybaras `w3c_click_offset` settomg
* Support setting color inputs

### Fixed

* No longer hangs starting the browser with JRuby
* ValidityState objects can now be returned from the browser
* Headers now returned correctly
* Drag type detection improved
* Mouse button status tracked correctly through actions
* Clicking on a link that splits across multiple lines now works
* Correctly handle JS exceptions that don't provide a stacktrace

# Version 0.3.0
Release date: 2019-05-28

### Added

* Node#obscured? now supports use in nested frames

### Changed

* Click is no longer used to focus most form fields to behave more like the selenium driver
* Removed backports gem requirement

# Version 0.2.0
Release date: 2019-05-17

### Added

* Windows chrome location detection
* Node#obscured? to support the new Capybara obscured filter
* Node#drop to support the new Capybara drop functionality currently in Capybara master

### Fixed

* Node#set passes options to set_text
* :return added as alias of :enter key
* correct signal passed to shutdown Chrome

# Version 0.1.0
Release date: 2019-02-05

### Status

* All planned Poltergeist tests passing


lib/capybara/apparition/driver.rb
# frozen_string_literal: true

require 'uri'
require 'forwardable'
require 'capybara/apparition/driver/chrome_client'
require 'capybara/apparition/configuration'
require 'capybara/apparition/browser/launcher'

module Capybara::Apparition
  class Driver < Capybara::Driver::Base
    DEFAULT_TIMEOUT = 30

    extend Forwardable

    attr_reader :app, :options

    delegate %i[restart current_url status_code body
                title frame_title frame_url switch_to_frame
                window_handles close_window switch_to_window
                paper_size= zoom_factor=
                scroll_to
                network_traffic clear_network_traffic
                headers headers= add_headers
                cookies all_cookies remove_cookie clear_cookies cookies_enabled=
                clear_memory_cache
                go_back go_forward refresh
                console_messages] => :browser

    def initialize(app, options = {})
      @app       = app
      @options   = options
      generate_browser_options
      @browser   = nil
      @inspector = nil
      @client    = nil
      @launcher  = nil
      @started   = false
    end

    def needs_server?
      true
    end

    # def chrome_url
    #   'ws://localhost:9223'
    # end

    def browser
      @browser ||= begin
        Browser.new(client, browser_logger) do |browser|
          browser.js_errors = options.fetch(:js_errors, true)
          browser.ignore_https_errors = options.fetch(:ignore_https_errors, false)
          browser.extensions = options.fetch(:extensions, [])
          browser.debug      = options.fetch(:debug, false)
          browser.url_blacklist = options[:url_blacklist] || []
          browser.url_whitelist = options[:url_whitelist] || []
        end
      end
    end

    def inspector
      @inspector ||= options[:inspector] && Inspector.new(options[:inspector])
    end

    def client
      @client ||= begin
        @launcher ||= Browser::Launcher.start(options)
        ws_url = @launcher.ws_url
        ::Capybara::Apparition::ChromeClient.client(ws_url.to_s)
      end
    end

    def quit
      @client&.stop
      @launcher&.stop
    end

    # logger should be an object that responds to puts, or nil
    def logger
      options[:logger] || (options[:debug] && STDERR)
    end

    # logger should be an object that behaves like IO or nil
    def browser_logger
      options.fetch(:browser_logger, $stdout)
    end

    def visit(url)
      @started = true
      browser.visit(url)
    end

    alias html body

    def source
      browser.source.to_s
    end

    def find(method, selector)
      browser.find(method, selector).map do |page_id, id, attrs|
        Capybara::Apparition::Node.new(self, page_id, id, attrs)
      end
    end

    def find_xpath(selector)
      find :xpath, selector.to_s
    end

    def find_css(selector)
      find :css, selector.to_s
    end

    def click(x, y)
      browser.click_coordinates(x, y)
    end

    def evaluate_script(script, *args)
      retry_if_wrong_world do
        unwrap_script_result(browser.evaluate(script, *native_args(args)))
      end
    end

    def evaluate_async_script(script, *args)
      retry_if_wrong_world do
        unwrap_script_result(browser.evaluate_async(script, session_wait_time, *native_args(args)))
      end
    end

    def execute_script(script, *args)
      retry_if_wrong_world do
        browser.execute(script, *native_args(args))
      end
      nil
    end

    def current_window_handle
      browser.current_window_handle
    end

    def no_such_window_error
      NoSuchWindowError
    end

    def reset!
      begin
        browser.reset
      rescue TimeoutError
        puts 'Reset timed out - retrying'
        browser.reset
      end
      browser.url_blacklist = options[:url_blacklist] || []
      browser.url_whitelist = options[:url_whitelist] || []
      @started = false
    end

    def save_screenshot(path, options = {})
      browser.render(path, options)
    end
    alias render save_screenshot

    def render_base64(format = :png, options = {})
      browser.render_base64(options.merge(format: format))
    end

    def resize(width, height)
      browser.resize(width, height, screen: options[:screen_size])
    end

    def resize_window(width, height)
      warn '[DEPRECATION] Capybara::Apparition::Driver#resize_window ' \
        'is deprecated. Please use Capybara::Window#resize_to instead.'
      resize(width, height)
    end

    def resize_window_to(handle, width, height)
      _within_window(handle) do
        resize(width, height)
      end
    end

    def maximize_window(handle)
      _within_window(handle) do
        browser.maximize
      end
    end

    def fullscreen_window(handle)
      _within_window(handle) do
        browser.fullscreen
      end
    end

    def window_size(handle)
      _within_window(handle) do
        evaluate_script('[window.innerWidth, window.innerHeight]')
      end
    end

    def set_proxy(host, port, type = nil, user_ = nil, password_ = nil, user: nil, password: nil, bypass: [])
      if user_ || password_
        warn '#set_proxy: Passing `user` and `password` as positional arguments is deprecated. ' \
             'Please pass as keyword arguments.'
        user ||= user_
        password ||= password_
      end

      # TODO: Look at implementing via the CDP Fetch domain when available
      @options[:browser_options] ||= {}
      @options[:browser_options]['proxy-server'] = "#{type + '=' if type}#{host}:#{port}"
      bypass = Array(bypass).join(';')
      @options[:browser_options]['proxy-bypass-list'] = bypass unless bypass.empty?
      browser.set_proxy_auth(user, password) if user || password
    end

    def add_header(name, value, options = {})
      browser.add_header({ name => value }, **{ permanent: true }.merge(options))
    end
    alias_method :header, :add_header

    def response_headers
      browser.response_headers.transform_keys { |key| key.split('-').map(&:capitalize).join('-') }
    end

    def set_cookie(name, value = nil, options = {})
      name, value, options = parse_raw_cookie(name) if value.nil?

      options[:name]  ||= name
      options[:value] ||= value
      options[:domain] ||= begin
        if @started
          URI.parse(browser.current_url).host
        else
          URI.parse(default_cookie_host).host || '127.0.0.1'
        end
      end

      browser.set_cookie(options)
    end

    def proxy_authorize(user = nil, password = nil)
      browser.set_proxy_auth(user, password)
    end

    def basic_authorize(user = nil, password = nil)
      browser.set_http_auth(user, password)
    end
    alias_method :authenticate, :basic_authorize

    def debug
      if @options[:inspector]
        # Fall back to default scheme
        scheme = begin
                   URI.parse(browser.current_url).scheme
                 rescue StandardError
                   nil
                 end
        scheme = 'http' if scheme != 'https'
        inspector.open(scheme)
        pause
      else
        raise Error, 'To use the remote debugging, you have to launch the driver ' \
                     'with `:inspector => true` configuration option'
      end
    end

    def pause
      # STDIN is not necessarily connected to a keyboard. It might even be closed.
      # So we need a method other than keypress to continue.

      # In jRuby - STDIN returns immediately from select
      # see https://github.com/jruby/jruby/issues/1783
      # TODO: This limitation is no longer true can we simplify?
      read, write = IO.pipe
      Thread.new do
        IO.copy_stream(STDIN, write)
        write.close
      end

      STDERR.puts "Apparition execution paused. Press enter (or run 'kill -CONT #{Process.pid}') to continue." # rubocop:disable Style/StderrPuts

      signal = false
      old_trap = trap('SIGCONT') do
        signal = true
        STDERR.puts "\nSignal SIGCONT received" # rubocop:disable Style/StderrPuts
      end
      # wait for data on STDIN or signal SIGCONT received
      keyboard = IO.select([read], nil, nil, 1) until keyboard || signal

      unless signal
        begin
          input = read.read_nonblock(80) # clear out the read buffer
          puts unless input&.end_with?("\n")
        rescue EOFError, IO::WaitReadable
          # Ignore problems reading from STDIN.
        end
      end
    ensure
      trap('SIGCONT', old_trap) # Restore the previous signal handler, if there was one.
      STDERR.puts 'Continuing' # rubocop:disable Style/StderrPuts
    end

    def wait?
      true
    end

    def invalid_element_errors
      [Capybara::Apparition::ObsoleteNode, Capybara::Apparition::MouseEventFailed, Capybara::Apparition::WrongWorld]
    end

    def accept_modal(type, options = {})
      case type
      when :alert
        browser.accept_alert
      when :confirm
        browser.accept_confirm
      when :prompt
        browser.accept_prompt options[:with]
      end

      yield if block_given?

      find_modal(options)
    end

    def dismiss_modal(type, options = {})
      case type
      when :confirm
        browser.dismiss_confirm
      when :prompt
        browser.dismiss_prompt
      end

      yield if block_given?
      find_modal(options)
    end

    def timeout
      client.timeout
    end

    def timeout=(sec)
      client.timeout = sec
    end

    def error_messages
      console_messages('error')
    end

    def within_window(selector, &block)
      warn 'Driver#within_window is deprecated, please switch to using Session#within_window instead.'
      _within_window(selector, &block)
      orig_window = current_window_handle
      switch_to_window(selector)
      begin
        yield
      ensure
        switch_to_window(orig_window)
      end
    end

    def version
      chrome_version = browser.command('Browser.getVersion')
      format(VERSION_STRING,
             capybara: Capybara::VERSION,
             apparition: Capybara::Apparition::VERSION,
             chrome: chrome_version['product'])
    end

    def open_new_window
      # needed because Capybara does arity detection on this method
      browser.open_new_window
    end

  private

    def retry_if_wrong_world
      timer = Capybara::Helpers.timer(expire_in: session_wait_time)
      begin
        yield
      rescue WrongWorld
        retry unless timer.expired?
        raise
      end
    end

    def _within_window(selector)
      orig_window = current_window_handle
      switch_to_window(selector)
      begin
        yield
      ensure
        switch_to_window(orig_window)
      end
    end

    def browser_options
      @options[:browser_options]
    end

    def generate_browser_options
      # TODO: configure SSL options
      # PhantomJS defaults to only using SSLv3, which since POODLE (Oct 2014)
      # many sites have dropped from their supported protocols (eg PayPal,
      # Braintree).
      # list += ["--ignore-ssl-errors=yes"] unless list.grep(/ignore-ssl-errors/).any?
      # list += ["--ssl-protocol=TLSv1"] unless list.grep(/ssl-protocol/).any?
      # list += ["--remote-debugger-port=#{inspector.port}", "--remote-debugger-autorun=yes"] if inspector
      # Note: Need to verify what Chrome command line options are valid for this
      browser_options = {}
      browser_options['remote-debugging-port'] = @options[:port] || 0
      browser_options['remote-debugging-address'] = @options[:host] if @options[:host]
      browser_options['window-size'] = @options[:window_size].join(',') if @options[:window_size]
      if @options[:browser]
        warn ':browser is deprecated, please pass as :browser_options instead.'
        browser_options.merge! process_browser_options(@options[:browser])
      end
      browser_options.merge! process_browser_options(@options[:browser_options]) if @options[:browser_options]
      if @options[:skip_image_loading]
        browser_options['blink-settings'] = [browser_options['blink-settings'], 'imagesEnabled=false'].compact.join(',')
      end

      @options[:browser_options] = browser_options
      process_cw_options(@options[:cw_options])
    end

    def process_cw_options(cw_options)
      return if cw_options.nil?

      (options[:url_blacklist] ||= []).concat cw_options[:url_blacklist]
      options[:js_errors] ||= cw_options[:js_errors]
    end

    def process_browser_options(options)
      case options
      when Array
        options.compact.each_with_object({}) do |option, hsh|
          if option.is_a? Hash
            hsh.merge! process_browser_options(option)
          else
            hsh[option.to_s.tr('_', '-')] = nil
          end
        end
      when Hash
        options.transform_keys { |option| option.to_s.tr('_', '-') }
      else
        raise ArgumentError, 'browser_options must be an Array or a Hash'
      end
    end

    def parse_raw_cookie(raw)
      parts = raw.split(/;\s*/)
      name, value = parts[0].split('=', 2)
      options = parts[1..-1].each_with_object({}) do |part, opts|
        name, value = part.split('=', 2)
        opts[name.to_sym] = value
      end
      [name, value, options]
    end

    def screen_size
      options[:screen_size] || [1366, 768]
    end

    def find_modal(options)
      timeout_sec   = options.fetch(:wait) { session_wait_time }
      expect_text   = options[:text]
      expect_regexp = expect_text.is_a?(Regexp) ? expect_text : Regexp.escape(expect_text.to_s)
      timer = Capybara::Helpers.timer(expire_in: timeout_sec)
      begin
        modal_text = browser.modal_message
        found_text ||= modal_text
        raise Capybara::ModalNotFound if modal_text.nil? || (expect_text && !modal_text.match(expect_regexp))
      rescue Capybara::ModalNotFound => e
        if timer.expired?
          raise e, 'Timed out waiting for modal dialog. Unable to find modal dialog.' unless found_text

          raise e, 'Unable to find modal dialog' \
                   "#{" with #{expect_text}" if expect_text}" \
                   "#{", did find modal with #{found_text}" if found_text}"
        end
        sleep(0.05)
        retry
      end
      modal_text
    end

    def session_wait_time
      if respond_to?(:session_options)
        session_options.default_max_wait_time
      else
        begin begin
                Capybara.default_max_wait_time
              rescue StandardError
                Capybara.default_wait_time
              end end
      end
    end

    def default_cookie_host
      if respond_to?(:session_options)
        session_options.app_host
      else
        Capybara.app_host
      end || ''
    end

    def native_args(args)
      args.map { |arg| arg.is_a?(Capybara::Apparition::Node) ? arg.native : arg }
    end

    def unwrap_script_result(arg, object_cache = {})
      return object_cache[arg] if object_cache.key? arg

      case arg
      when Array
        object_cache[arg] = []
        object_cache[arg].replace(arg.map { |e| unwrap_script_result(e, object_cache) })
        object_cache[arg]
      when Hash
        if (arg['subtype'] == 'node') && arg['objectId']
          tag_name = arg['description'].split(/[.#]/, 2)[0]
          Capybara::Apparition::Node.new(self, browser.current_page, arg['objectId'], tag_name: tag_name)
        else
          object_cache[arg] = {}
          arg.each { |k, v| object_cache[arg][k] = unwrap_script_result(v, object_cache) }
          object_cache[arg]
        end
      else
        arg
      end
    end

    VERSION_STRING = <<~VERSION
      Versions in use:
      Capybara: %<capybara>s
      Apparition: %<apparition>s
      Chrome: %<chrome>s
    VERSION
  end
end
